Climate Index Implementation

antoinehermant opened the pull request:
Implementation of a Climate Index approach for glacial/interglacial simulations.

For the atmosphere, a standalone model computes timeseries of air temperature and precipitation.
For the ocean, the climate index is incorporated into PICO, providing it with timeseries of theta and salinity.

**Checklist**

- [x] update documentation
- [x] update [`CHANGES.rst`](CHANGES.rst)
- [ ] add regression tests

antoinehermant committed:
Whitespace cleanup

antoinehermant committed:
Climate Index Implementation

Implementation of a Climate Climate Index timeseries for glacial/interglacial simulations.
The index is based on Ice-Core records of the climate state (see Sutter et al. 2019) and is used to scale forcing snapshots between the time slices.

For the atmosphere, the index is used to scale anomaly snapshots of the air temperature and precipitation.
The precipitation can also be scaled using either an uniform or spacial scaling factor (read from an input file).
In addition, the air temperature uses a cosinus approach using annual and mean summer temperatures for representing the yearly cycle (copied from YearlyCycle).

For the ocean, the climate index is implemented into PICO, and is used to scale anomaly snapshots of the potential temperature and salinity.

ckhroulev committed:
Silence overly verbose messages

ckhroulev committed:
atmosphere::ClimateIndex: fix begin_access()/end_access() usage

It's best to use array::AccessScope instead.

ckhroulev committed:
atmosphere::ClimateIndex: remove m_w0, m_w1, m_w1X

ckhroulev committed:
Clean up ClimateIndexWeights

- Remove m_current, rename m_W0, m_W1, m_W1X (remove "m_")
- update_weights(t, dt) returns std::array<double,3> instead of using output arguments

ckhroulev committed:
Clean up ocean::ClimateIndex

ckhroulev committed:
ocean::Pico::use_ci is not needed

ckhroulev committed:
Clean up atmosphere::ClimateIndex

ckhroulev commented:
I cleaned up the code a bit -- please take a look and confirm that it still does what it should.

Here are some comments about aspects of the code and documentation that need more attention:

## Initialization

The initialization code (`ocean::ClimateIndex::init_forcing()` and `atmosphere::ClimateIndex::init_impl()` needs more work. In particular, we should not decide what to do based on what is available in the file (e.g. whether to use "1X" anomalies). A user should select a particular configuration (with or without "1X", with or without temperature-anomaly-based precipitation scaling, etc) and the code should stop with an error message if required fields are not available.

## Time-dependence of precipitation

When both the cosine annual cycle for air temperature *and* the temperature-driven precipitation scaling are enabled, the reported 2D "mean precipitation" **is** scaled using "annual" temperature anomalies, but values of precipitation time series may need to be scaled using temperature time series resolving the annual cycle. We should either change the code so that the cosine annual cycle is used to scale precipitation time series, *or* document that this is not done.

## Variable metadata

Several variables have incorrect metadata, i.e. descriptions ("long name") and units -- both in the code and documentation.

## Documentation

### Air temperature

It would be better to state the scaling formula (`air_temp_annual = ...`) in the style of the formulas 3 and 4 in Sutter et al, 2019. The correspondence between quantities in this formula and variables can be described in a table.

Also: formula 3 in Sutter et al says

$$T_s(t) = T_{spd} + \sum_{x=ig,g,p} w_x(t) \times \Delta T_x$$

where $\Delta T_x$ are "the climate anomalies for the LIG, the LGM and the Pliocene, respectively". It is also stated that

> For each climate snapshot, anomaly fields with respect to the pre-industrial control run are calculated and added to a mean Antarctic climatology (1979–2011), derived from the regional climate model RACMO ([van Wessem et al.](https://tc.copernicus.org/articles/13/2023/2019/#bib1.bibx56), [2014](https://tc.copernicus.org/articles/13/2023/2019/#bib1.bibx56)), or the extrapolated World Ocean Atlas 2009 ([Locarnini et al.](https://tc.copernicus.org/articles/13/2023/2019/#bib1.bibx31), [2010](https://tc.copernicus.org/articles/13/2023/2019/#bib1.bibx31)) to provide the climate forcing for the individual climate epoch.

Note that here, if I understand correctly, **all** $\Delta T_x$ are computed "with respect to the pre-industrial control run."

I don't understand why your code uses a different factor (`(m_air_temp_anomaly_annual_1X(i, j) - m_air_temp_anomaly_annual_1(i, j))`) with the `m_w1X` weight. Could you please explain this to me? Thanks!

### References

We need to cite *J. Sutter et al. Modelling the Antarctic Ice Sheet across the mid-Pleistocene transition – implications for Oldest Ice. The Cryosphere, 13(7):2023–2041, 2019. [doi:10.5194/tc-13-2023-2019](https://doi.org/10.5194/tc-13-2023-2019)* and any other relevant papers. It should be clear which formulas from what sections of cited paper(s) are implemented. If implemented formulas do not correspond to anything in published literature we need to explain what is different.

### Inputs

We should explain where one would get required input data from (variables `climate_index`, `air_temp_annual_ref`, `precipitation_ref`, `air_temp_anomaly_annual_0`, `precipitation_anomaly_0`, etc) and cite relevant papers and data sources. In particular, we should document how to get required "anomalies" from climate snapshots.

### Testing

We need to create a setup for testing this code. It's okay if it uses a very coarse grid and dummy inputs, but it should use the code in a non-trivial fashion, covering as many different configurations as practicable (with and without "1X" anomalies, with and without the cosine annual cycle, with and without precipitation scaling). We can probably use coarse grid datasets we use to test PICO.

## Authorship

Please add people who helped develop this code to `CITATION.cff` so that we can properly acknowledge their contributions.

antoinehermant commented:
Thank you very much Constantine for your inputs.

We are involved in reviewing a community paper this week, we will take time
next week to implement the changes you are suggesting and clarify the
formulas in the paper.

Cheers,
Antoine

On Thu, 15 Feb 2024 at 00:27, Constantine Khrulev ***@***.***>
wrote:

> I cleaned up the code a bit -- please take a look and confirm that it
> still does what it should.
>
> Here are some comments about aspects of the code and documentation that
> need more attention:
> Initialization
>
> The initialization code (ocean::ClimateIndex::init_forcing() and
> atmosphere::ClimateIndex::init_impl() needs more work. In particular, we
> should not decide what to do based on what is available in the file (e.g.
> whether to use "1X" anomalies). A user should select a particular
> configuration (with or without "1X", with or without
> temperature-anomaly-based precipitation scaling, etc) and the code should
> stop with an error message if required fields are not available.
> Time-dependence of precipitation
>
> When both the cosine annual cycle for air temperature *and* the
> temperature-driven precipitation scaling are enabled, the reported 2D "mean
> precipitation" *is* scaled using "annual" temperature anomalies, but
> values of precipitation time series may need to be scaled using temperature
> time series resolving the annual cycle. We should either change the code so
> that the cosine annual cycle is used to scale precipitation time series,
> *or* document that this is not done.
> Variable metadata
>
> Several variables have incorrect metadata, i.e. descriptions ("long name")
> and units -- both in the code and documentation.
> Documentation Air temperature
>
> It would be better to state the scaling formula (air_temp_annual = ...)
> in the style of the formulas 3 and 4 in Sutter et al, 2019. The
> correspondence between quantities in this formula and variables can be
> described in a table.
>
> Also: formula 3 in Sutter et al says
>
> $$T_s(t) = T_{spd} + \sum_{x=ig,g,p} w_x(t) \times \Delta T_x$$
>
> where $\Delta T_x$ are "the climate anomalies for the LIG, the LGM and
> the Pliocene, respectively". It is also stated that
>
> For each climate snapshot, anomaly fields with respect to the
> pre-industrial control run are calculated and added to a mean Antarctic
> climatology (1979–2011), derived from the regional climate model RACMO (van
> Wessem et al.
> <https://tc.copernicus.org/articles/13/2023/2019/#bib1.bibx56>, 2014
> <https://tc.copernicus.org/articles/13/2023/2019/#bib1.bibx56>), or the
> extrapolated World Ocean Atlas 2009 (Locarnini et al.
> <https://tc.copernicus.org/articles/13/2023/2019/#bib1.bibx31>, 2010
> <https://tc.copernicus.org/articles/13/2023/2019/#bib1.bibx31>) to
> provide the climate forcing for the individual climate epoch.
>
> Note that here, if I understand correctly, *all* $\Delta T_x$ are
> computed "with respect to the pre-industrial control run."
>
> I don't understand why your code uses a different factor ((m_air_temp_anomaly_annual_1X(i,
> j) - m_air_temp_anomaly_annual_1(i, j))) with the m_w1X weight. Could you
> please explain this to me? Thanks!
> References
>
> We need to cite *J. Sutter et al. Modelling the Antarctic Ice Sheet
> across the mid-Pleistocene transition – implications for Oldest Ice. The
> Cryosphere, 13(7):2023–2041, 2019. doi:10.5194/tc-13-2023-2019
> <https://doi.org/10.5194/tc-13-2023-2019>* and any other relevant papers.
> It should be clear which formulas from what sections of cited paper(s) are
> implemented. If implemented formulas do not correspond to anything in
> published literature we need to explain what is different.
> Inputs
>
> We should explain where one would get required input data from (variables
> climate_index, air_temp_annual_ref, precipitation_ref,
> air_temp_anomaly_annual_0, precipitation_anomaly_0, etc) and cite
> relevant papers and data sources. In particular, we should document how to
> get required "anomalies" from climate snapshots.
> Testing
>
> We need to create a setup for testing this code. It's okay if it uses a
> very coarse grid and dummy inputs, but it should use the code in a
> non-trivial fashion, covering as many different configurations as
> practicable (with and without "1X" anomalies, with and without the cosine
> annual cycle, with and without precipitation scaling). We can probably use
> coarse grid datasets we use to test PICO.
> Authorship
>
> Please add people who helped develop this code to CITATION.cff so that we
> can properly acknowledge their contributions.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pism/pism/pull/534#issuecomment-1944999634>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUUWTW7GV2RZZZQRHJENNHTYTVB5XAVCNFSM6AAAAABDFH45E6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNBUHE4TSNRTGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

antoinehermant committed:
Clean up Climate Index configuration:

- flag to use cosinus for yearly cycle: -use_cosinus_yearly_cycle
- flag to use super interglacial snapshot: -use_super_interglacial

Added cosinus yeary cycle for precipitation scaling temperature difference between annual and summer mean
Use with flag -use_precip_cosinus_yearly_cycle

antoinehermant committed:
prepare to merge with pism2.2
